pkg/scheduler/util/error_channel.go

```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import "context"
+
+// ErrorChannel supports non-blocking send and receive operation to capture error.
+type ErrorChannel struct {
+	errCh chan error
+}
+
+// SendError sends an error without blocking the receiver.
+func (e *ErrorChannel) SendError(err error) {
+	select {
+	case e.errCh <- err:
+	default:
+	}
+}
+
+// SendErrorWithCancel sends an error without blocking the receiver and calls
+// cancel function.
+func (e *ErrorChannel) SendErrorWithCancel(err error, cancel context.CancelFunc) {
+	e.SendError(err)
+	cancel()
+}
+
+// ReceiveError receives a error from channel without blocking on the sender.
+func (e *ErrorChannel) ReceiveError() error {
+	select {
+	case err := <-e.errCh:
+		return err
+
+	default:
+		return nil
+	}
+}
+
+// NewErrorChannel returns a new ErrorChannel.
+func NewErrorChannel(bufferSize int64) *ErrorChannel {
```

Comment:
We shouldn't allow passing a zero here. If by mistake someone created an unbuffered channel,  ReceiveError may potentially miss reading the error, and so will return nil even though an error was sent.

I suggest we don't cater for the case of collecting multiple errors and just fix the buffer size to 1 (ie don't allow passing the buffer size as a parameter) to simplify the interface and make it less prone to misuse.

Reply:
Done